### ISSUE TITLE
Debugging with new TS async/await

### ISSUE THREAD

**Issue by wallverb** [Microsoft/vscode#11645]
- VSCode Version: Code - Insiders 1.5.0-insider (b2c874b253732778a246b94aa1eba93806c45361, 2016-09-07T11:56:57.966Z)
- OS Version: Darwin x64 15.6.0

Since we got an early xmass gift from TypeScript team today with ES5 async/await it would be great if VSCode could better handle it. Right now running the example and pressing F10 to step over `await` takes us to js file

Steps to Reproduce:

```ts
"use strict";

// printDelayed is a 'Promise<void>'
async function printDelayed(elements: string[]) {
    for (const element of elements) {
        await delay(200);
        console.log(element);
    }
}

async function delay(milliseconds: number) {
    return new Promise<void>(resolve => {
        setTimeout(resolve, milliseconds);
    });
}

printDelayed(["Hello", "beautiful", "asynchronous", "world"]).then(() => {
    console.log();
    console.log("Printed every element!");
});
```

Put breakpoint as the following:

![image](https://cloud.githubusercontent.com/assets/1723815/18323499/df8d3f82-7505-11e6-8a90-ef20301dceba.png)

Hoping to be able to press `F10` to get to the next line but user is taken here:

![image](https://cloud.githubusercontent.com/assets/1723815/18323514/f3aceb84-7505-11e6-8c7d-f3bfb86fb0ec.png)

Seems like ES6 target also is not working properly, user is taken here with `F10`:

![image](https://cloud.githubusercontent.com/assets/1723815/18323583/3e9f5b04-7506-11e6-8132-377960a2375c.png)

**Comment by wallverb** [comment]
Related: https://github.com/Microsoft/vscode/issues/9368

**Comment by Phmager** [comment]
+1 we need better async debugging support

**Comment by roblourens** [comment]
Well, that is the next line that actually executes. Chrome devtools does the same thing. I don't think we can improve this since the debugger doesn't understand your code.

When debugging with async/await and TS, I find it really useful to set `smartStep` in my launch config. This will make the debugger auto-step over the JS stuff that you end up stepping into all the time.

**Comment by weinand** [comment]
Here is some `smartStep` doc: https://code.visualstudio.com/docs/editor/node-debugging#_smart-stepping

**Comment by chadbr** [comment]
"smartStep": true makes a big difference - but it could still be better?

I'd love it to be at the level of C# debugging of async/await

**Comment by weinand** [comment]
@chadbr reaching the  C# level of async/await debugging is difficult for a transpiled language because the runtime only sees the generated JavaScript and the source maps have there limitations.

**Comment by chadbr** [comment]
The debugger doesn't have access to the same information the editors have for navigation / intelli-sense, etc.?

**Comment by weinand** [comment]
@chadbr yes, the v8-debugger used in node has no access to the language server (and is not under our control). So it does not even understand source maps. 
The VS Code debug extension adds source map support on top of the debugger API. But this support is still language agnostic (and there are lots of languages that are transpiled to JavaScript).

**Comment by chadbr** [comment]
gotcha -- need a new debugger in V8 :)

**Comment by polkovnikov-ph** [comment]
@weinand It's possible to make v8 debugger do steps until the next "relevant" instruction is met. There's a similar feature in IE that allows to omit steps in library code.

**Comment by weinand** [comment]
@polkovnikov-ph the "smartStep" and "skipFiles" features can be used to automatically step over "uninteresting" code. See https://code.visualstudio.com/docs/editor/node-debugging#_smart-stepping

**Comment by polkovnikov-ph** [comment]
@weinand Yes, that's exactly what I wrote :) Thanks for linking the manual, it will be interesting for future readers.

**Comment by weinand** [comment]
@polkovnikov-ph aha, so with "v8 debugger" you meant the VS Code node-debugger. Sorry, I thought you referred to something else.

**Comment by chadbr** [comment]
fwiw, I have had great luck switching to es6 for debugging -- and switching back for testing / deploy.

**Comment by insprintorob** [comment]
I also had luck experimenting with higher ES versions for the compilation target.

es6 works fine for me with the latest node 7 and 8 releases.

**Comment by Deilan** [comment]
I also experience this issue a lot. `smartStep` and `skipFiles` don't help because stepping out a line of code with `await` results in forwarding execution context to the end of the enclosing method within the current typescript file. So there should be more general and consistent solution.

P. S. The only workaround I've found is to set a breakpoint at the next line and continue execution (<kbd>F5</kbd>) till it hits the previously set breakpoint.

**Comment by Deilan** [comment]
Somewhat related with #30382.

**Comment by OliverJAsh** [comment]
Does this issue also cover the same behaviour but for standard (non-async/await) Promise code? E.g.

``` ts
/*paused here, step into… */Promise.resolve(1)
.then(() => {
  /* step into lands here */
  /* step into again… */
})
.then(() => {
  /* step into lands here */
})
```

I tried enabling `smartStep`, but currently I get this behaviour:

``` ts
/*paused here, step into… */Promise.resolve(1)
.then(() => {

})
/* step into lands here */
/* step into again… */
.then(() => {

})
/* step into lands here */
```

**Comment by zwhitchcox** [comment]
> fwiw, I have had great luck switching to es6 for debugging -- and switching back to es5 for testing / deploy.

You are a genius sir, thank you

**Comment by zzxoto** [comment]
Here are my settings, if this illuminates others. Thanks to @zwhitchcox @chadbr for inspiration.

launch.json file looks like:
```json
{
      "type": "node",
      "request": "attach",
      "name": "Attach by Process ID",
      "processId": "${command:PickProcess}",
      "outFiles": [ 
        "${workspaceFolder}/dist/*.js",
        "${workspaceFolder}/dist/**/*.js",
       ],
       "smartStep": true,
       "skipFiles": [
        "${workspaceFolder}/node_modules/**/*.js",
        "${workspaceFolder}/node_modules/**/**/*.js",
        "${workspaceFolder}/node_modules/**/**/**/*.js",
        "${workspaceFolder}/node_modules/**/**/**/**/*.js",
       ],
      "sourceMaps": true,
      "protocol": "inspector",
    },
```
Note: vscode documentations recommends
```
  "skipFiles": [
     "<node_internals>/**/*.js"
   ]
```
but doing so, skips aggressively i found.

tsconfig.json should have

```json

    //NOTE: poor debugger support for lower versions than this
    //by poor I mean, skipping to .js files too often and too deep
    "target": "es2017",

    //since target is es2017, why not this eh! Haven't researched this enough
    "lib": ["es2017"],
    
    //this makes source map attach at the end of file rather than spamming
   //seperate *.map file. This should not have any affect on debugger jumping though
   //having this options as true, means the sourceMap option should not be present
    "inlineSourceMap": true,
```

Running vscode 1.28.2 with typescript 3.5.2
Node.js 10.16.0
